src/rgw/rgw_rados.cc

```diff
@@ -5637,7 +5637,7 @@ int RGWRados::Object::Read::prepare(optional_yield y)
       string if_nomatch_str = rgw_string_unquote(conds.if_nomatch);
       ldout(cct, 10) << "ETag: " << string(etag.c_str(), etag.length()) << " " << " If-NoMatch: " << if_nomatch_str << dendl;
       if (if_nomatch_str.compare(0, etag.length(), etag.c_str(), etag.length()) == 0) {
-        return -ERR_NOT_MODIFIED;
+        return -ERR_PRECONDITION_FAILED;
```

Comment:
this change looks correct for the CopyObj case, but not for the general case of GetObj with If-None-Match according to https://tools.ietf.org/html/rfc7232#section-3.2:
>    An origin server MUST NOT perform the requested method if the
   condition evaluates to false; instead, the origin server MUST respond
   with either a) the 304 (Not Modified) status code if the request
   method is GET or HEAD or b) the 412 (Precondition Failed) status code
   for all other request methods.

i feel like it's appropriate for `RGWRados::Object::Read::prepare()` to return `-ERR_NOT_MODIFIED` here so the GetObj case continues to work, but `RGWRados::copy_obj()` should map this `-ERR_NOT_MODIFIED` error to `-ERR_PRECONDITION_FAILED`

what do you think?

Reply:
thanks, @cbodley  you are right, i have changed it.

Reply:
@cbodley Please review it for me. Thanks